### PR TITLE
Added taint tool window and menu command

### LINKED ISSUE #1979: Add Taint tool window
#### Description

- [x] Create a new tool window for the taint vulnerabilities control.
- [x] Create menu command to display the window

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Fixes #1979

### CODE REVIEW

File: src/IssueViz.Security.UnitTests/Commands/HotspotsToolWindowCommandTests.cs [hunk 1]
```diff
@@ -66,8 +67,8 @@ public void Ctor_CommandAddedToMenu()
 
             commandService.Verify(x =>
                     x.AddCommand(It.Is((MenuCommand c) =>
-                        c.CommandID.Guid == HotspotsToolWindowCommand.CommandSet &&
-                        c.CommandID.ID == HotspotsToolWindowCommand.ViewToolWindowCommandId)),
+                        c.CommandID.Guid == Constants.CommandSetGuid &&
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Commands/HotspotsToolWindowCommandTests.cs:70]: Minor refactoring. All of the command constants are in a single class and not duplicated in other classes.

File: src/IssueViz.Security.UnitTests/Commands/TaintToolWindowCommandTests.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.Integration.UnitTests;
+using SonarLint.VisualStudio.IssueVisualization.Security.Commands;
+using SonarLint.VisualStudio.IssueVisualization.Security.UI.TaintList;
+using Constants = SonarLint.VisualStudio.IssueVisualization.Security.Commands.Constants;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.Commands
+{
+    [TestClass]
+    public class TaintToolWindowCommandTests
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Commands/TaintToolWindowCommandTests.cs:36]: Copy of the corresponding hotspots tests.

Looking at the implementation of the command class, I think it should be possible to refactor it to a single general purpose "show tool window" command implementation that is just passed the [command guid+id] and [toolwindow guid]. I'll look that separately.

The `IssueVisualizationToolWindowCommand` is similar but slightly more complicated (it registers two commands and sets the enabled status based an active UI context). I'll ignore it for now, but it's something we might want to add to the security "show tool window" commands in the future e.g. the show hotspots window command only be enabled if there is a hotspot in the list.

File: src/IssueViz.Security.UnitTests/TaintList/TaintToolWindowTests.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+using Moq;
+using SonarLint.VisualStudio.IssueVisualization.Security.UI.TaintList;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.UnitTests.TaintList
+{
+    [TestClass]
+    public class TaintToolWindowTests
+    {
+        [TestMethod]
+        public void Dispose_FrameIsClosed()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/TaintList/TaintToolWindowTests.cs:33]: Also a copy of the hotspot tests.

File: src/IssueViz.Security/Commands/Constants.cs [hunk 4]
```diff
@@ -25,5 +25,8 @@ namespace SonarLint.VisualStudio.IssueVisualization.Security.Commands
     internal static class Constants
     {
         public static readonly Guid CommandSetGuid = new Guid("97856422-20A2-4DB5-A468-1BAA9B6EEC38");
+
+        public const int HotspotsToolWindowCommandId = 0x0100;
+        public const int TaintToolWindowCommandId = 0x0101;
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Commands/Constants.cs:30]: Now that we have multiple commands it's convenient to have the id constants in the same file - makes it easier to check that we are using the same values as in the `vsct`.

File: src/IssueViz.Security/Commands/HotspotsToolWindowCommand.cs [hunk 5]
```diff
@@ -31,9 +31,6 @@ namespace SonarLint.VisualStudio.IssueVisualization.Security.Commands
 {
     internal sealed class HotspotsToolWindowCommand
     {
-        public static readonly Guid CommandSet = Constants.CommandSetGuid;
-        public const int ViewToolWindowCommandId = 0x0100;
-
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Commands/HotspotsToolWindowCommand.cs:36]: All the command constants are now in a single file.

File: src/IssueViz.Security/Commands/IssueVizSecurityPackage.vsct [hunk 6]
```diff
@@ -16,6 +16,14 @@
         </Strings>
       </Button>
 
+      <Button guid="guidHotspotsCmdSet" id="TaintToolWindowCommandId" priority="0x0101" type="Button">
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Commands/IssueVizSecurityPackage.vsct:19]: Copy of the existing command with the priority incremented by one so it appears after it in the menu.

File: src/IssueViz.Security/Commands/TaintToolWindowCommand.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Shell;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.IssueVisualization.Security.UI.TaintList;
+using Task = System.Threading.Tasks.Task;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.Commands
+{
+    internal sealed class TaintToolWindowCommand
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Commands/TaintToolWindowCommand.cs:32]: Boilerplate.

File: src/IssueViz.Security/Resources.resx [hunk 8]
```diff
@@ -118,9 +118,15 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="ERR_HotspotsToolWindow_Exception" xml:space="preserve">
-    <value>Error displaying Hotspots Tool Window: {0}</value>
+    <value>Error displaying Security Hotspots Tool Window: {0}</value>
```
Comment (duncanp-sonar) [inline src/IssueViz.Security/Resources.resx:121]: Miscellaneous - changed `Hotspots` to `Security Hotspots`

File: src/IssueViz.Security/Commands/TaintToolWindowCommand.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            try
+            {
+                toolWindowService.Show(TaintToolWindow.ToolWindowId);
+            }
+            catch (Exception ex) when (!ErrorHandler.IsCriticalException(ex))
+            {
+                logger.WriteLine(string.Format(Resources.ERR_HotspotsToolWindow_Exception, ex));
```
Comment (rita-g-sonarsource) [inline src/IssueViz.Security/Commands/TaintToolWindowCommand.cs:79]: Wrong error message

Comment (rita-g-sonarsource) [inline src/IssueViz.Security/Commands/TaintToolWindowCommand.cs:79]: Wrong error message

File: src/IssueViz.Security/UI/TaintList/TaintVulnerabilitiesControl.xaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+    <UserControl.Resources>
+        <ResourceDictionary>
+            <ResourceDictionary.MergedDictionaries>
+                <ResourceDictionary Source="..\SharedResources.xaml" />
+            </ResourceDictionary.MergedDictionaries>
+        </ResourceDictionary>
+    </UserControl.Resources>
+    <Grid UseLayoutRounding="True">
+        <TextBlock Text="Under construction" />
```
Comment (rita-g-sonarsource) [inline src/IssueViz.Security/UI/TaintList/TaintVulnerabilitiesControl.xaml:13]: Where is the 90s' banner? 😜

File: src/IssueViz.Security.UnitTests/Commands/TaintToolWindowCommandTests.cs (hunk 2 above)
Comment (rita-g-sonarsource) [reply src/IssueViz.Security.UnitTests/Commands/TaintToolWindowCommandTests.cs:36]: I agree, good idea. The boilerplate is identical.

**Review by rita-g-sonarsource — APPROVED** [review]

**Pull request merged**